apps/emqx/src/emqx_channel.erl

```diff
@@ -2188,15 +2198,152 @@ ensure_disconnected(
 
 %%--------------------------------------------------------------------
 %% Maybe Publish will msg
-
-maybe_publish_will_msg(Channel = #channel{will_msg = undefined}) ->
+%% @doc May publish will message [MQTT-3.1.2-8]
+%%   When willmsg presents the decision whether or when to publish the Will Message are effected by
+%%   the followings:
+%%   - connecion state
+%%   - If it is MQTT normal disconnection (RC: 0)
+%%   - If it is MQTT normal disconnection (RC: 4)
+%%   - will delay interval (MQTT 5.0 only)
+%%   - session expire Session Expiry (MQTT 5.0 only)
+%%   - EMQX operations on the client
+%% @NOTE:
+%%    Connection close with session expiry interval = 0 means session close.
+%% @NOTE:
+%%    The caller does not need to take care of the case when process terminates while will_msg is published
+%%    as it is designed by the spec.
+-spec maybe_publish_will_msg(Reason, channel()) -> channel() when
+    Reason ::
+        %% Connection is terminating because session is taken over by another process.
+        takenover
+        %% Connection is terminating because of EMQX mgmt operation, also delete the session.
+        | kicked
+        %% Connection is terminating because session is taken over by another process.
+        | discarded
+        %% Connection is terminating because session is expired
+        | expired
+        %% Connection is terminating because of socket close/error
+        | sock_closed
+        %% Connection is terminating with Reasons
+        | {shutdown, atom()}
+        %% Connection is terminating, delay willmsg publish is impossible.
+        | ?chan_terminating
+        %% Connection is terminating because of normal MQTT disconnection, implies delete the session.
+        | normal.
+maybe_publish_will_msg(normal, Channel) ->
+    %% [MQTT-3.1.2-8]
     Channel;
-maybe_publish_will_msg(Channel = #channel{clientinfo = ClientInfo, will_msg = WillMsg}) ->
+maybe_publish_will_msg(_Reason, Channel = #channel{will_msg = undefined}) ->
+    %% No will message to publish
+    Channel;
+maybe_publish_will_msg(
+    _Reason,
+    Channel = #channel{
+        conn_state = ConnState,
+        conninfo = #{clientid := ClientId}
+    }
+) when
+    ConnState =:= idle orelse
+        ConnState =:= connecting orelse
+        ConnState =:= reauthenticating
+->
+    %% Wrong state to publish
+    ?tp(debug, willmsg_wrong_state, #{clientid => ClientId}),
```

Comment:
I think EMQX should also send the will message if re-authentication fails causing the connection to be closed

Reply:
I don't agree.
EMQX should never publish a will for unauthenticated or unauthorised clients.

Reply:
@zmstone I think we meant the reauth after conencted (authenticated and authorised) because session state won't be updated with willmsg if auth during connect is failed. 

according to the spec [MQTT-3.1.2-8] line 617-618, if the client connection isn't closed with Reason 0 (NOERROR), server should try to publish willmsg.